library/core/src/pin.rs

```diff
@@ -1,188 +1,621 @@
-//! replacing and moving the values they contain: you can move out of a <code>[Box]\<T></code>,
-//! or you can use [`mem::swap`]. <code>[Pin]\<P></code> wraps a pointer type `P`, so
-//! <code>[Pin]<[Box]\<T>></code> functions much like a regular <code>[Box]\<T></code>:
-//! when a <code>[Pin]<[Box]\<T>></code> gets dropped, so do its contents, and the memory gets
-//! deallocated. Similarly, <code>[Pin]<[&mut] T></code> is a lot like <code>[&mut] T</code>.
-//! However, <code>[Pin]\<P></code> does not let clients actually obtain a <code>[Box]\<T></code>
-//! or <code>[&mut] T</code> to pinned data, which implies that you cannot use operations such
-//! as [`mem::swap`]:
+//! Types that pin data to a location in memory.
+//!
+//! It is sometimes useful to be able to rely upon a certain value not being able to *move*,
+//! in the sense that its address in memory cannot change. This is useful especially when there
+//! are one or more [*pointers*][pointer] pointing at that value. The ability to rely on this
+//! guarantee that the value a [pointer] is pointing at (its **pointee**) will
+//!
+//! 1. Not be *moved* out of its memory location
+//! 2. More generally, remain *valid* at that same memory location
+//!
+//! is called "pinning." We would say that a value which satisfies these guarantees has been
+//! "pinned," in that it has been permanently (until the end of its lifetime) attached to its
+//! location in memory, as though pinned to a pinboard. Pinning a value is incredibly useful in
+//! that it provides the necessary guarantees[^guarantees] for [`unsafe`] code to be able to
+//! dereference raw pointers to the pinned value for the duration it is pinned (which,
+//! [as we'll see later][drop-guarantee], is necessarily from the time the value is first pinned
+//! until the end of its lifetime). This concept of "pinning" is necessary to implement safe
+//! interfaces on top of things like self-referential types and intrusive data structures which
+//! cannot currently be modeled in fully safe Rust using only borrow-checked
+//! [references][reference].
+//!
+//! "Pinning" allows us to put a *value* which exists at some location in memory into a state where
+//! safe code cannot *move* that value to a different location in memory or otherwise invalidate it
+//! at its current location (unless it implements [`Unpin`], which we will
+//! [talk about below][self#unpin]). Anything that wants to interact with the pinned value in a way
+//! that has the potential to violate these guarantees must promise that it will not actually
+//! violate them, using the [`unsafe`] keyword to mark that such a promise is upheld by the user
+//! and not the compiler. In this way, we can allow other [`unsafe`] code to rely on any pointers
+//! that point to the pinned value to be valid to dereference while it is pinned.
+//!
+//! [^guarantees]: Pinning on its own does not provide *all* the invariants necessary here. However,
```

Comment:
I'm confused by this footnote. Which invariants are missing?

Reply:
What I meant to express is that pinning's invariants *add* to the standard Rust invariants, i.e. that an owned value must always be valid and a reference must always be to a valid value. Pinning does not *itself* guarantee those, but rather assumes they are already true and then adds its own extra invariants on top. Not sure how to express that better/more concisely but if you have suggestions on this especially and whther you think it's even worth clarifying, I'd love to hear them.

Reply:
Ah I see, thanks for explaining.

My first reaction is that we are better off without the footnote. I find it unlikely that someone would assume that pinning takes care of pointer dereferencability for free.

Maybe we can weaken the text where this footnote is linked to make this less likely. Currently it says
> Pinning a value is incredibly useful in that it provides the necessary guarantees[^guarantees] for [`unsafe`] code to be able to dereference raw pointers to the pinned value for the duration it is pinned.

What about something like
"Pinning a value is an incredibly useful building block for [`unsafe`] code to be able to reason about whether a raw pointer to the pinned value is still valid."

Reply:
Yeah, this is good, applied. With the new text I think the footnote is a bit of a distraction; it may be worth preserving that language for elsewhere but this section doesn't need to go into that much depth imo.